### PR TITLE
[Form] fix problems when the field values are Boolean

### PR DESCRIPTION
**Pull request by qlombat**
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | - |
| License | MIT |
| Doc PR | - |
# Problem

When the field values are Boolean as below, false value is not shown.
class RadioYesNoType extends AbstractType

```
{
    public function setDefaultOptions(OptionsResolverInterface $resolver)
    {
        $resolver->setDefaults(array(
            'choices' => array(
                false => 'Yes',
                true => 'No',
            ),
            'multiple'  => false,
            'expanded'  => true,
        ));
    }

    public function getParent()
    {
        return 'choice';
    }

    public function getName()
    {
        return 'Radio_Yes_No';
    }
}
```

### CODE REVIEW

**Comment by javiereguiluz** [comment]
@be-bewweb please, update the description of your pull request to add the checklist as explained here: http://symfony.com/doc/current/contributing/code/patches.html#make-a-pull-request Thanks!

**Comment by xabbuh** [comment]
If this fixes a bug, please also add a test to avoid future regressions.

**Comment by be-bewweb** [comment]
Where can I add the test ? Sorry but this is the first time I do a pull request. Thanks for your help

**Comment by fabpot** [comment]
@be-bewweb The tests for the Form component are stored under the `src/Symfony/Component/Form/Tests` directory. The tests for the Form class are in the `src/Symfony/Component/Form/Tests/SimpleFormTest.php`, which is regular PHPUnit test class. You should add a new test in this file to cover your fix so that we can ensure that we won't introduce regressions in the future.

**Comment by stof** [comment]
@be-bewweb the issue is that choices in your case are not ``true`` and ``false`` but ``'1'`` and ``''``, because booleans are not valid array keys and so PHP casts them to strings when building the array.

So IMO, this PR is invalid.
The issue is in your own code, where the actual value is not the one you think.

File: src/Symfony/Component/Form/Form.php [hunk 1]
```diff
@@ -344,7 +344,7 @@ public function setData($modelData)
         }
 
         // Treat data as strings unless a value transformer exists
-        if (!$this->config->getViewTransformers() && !$this->config->getModelTransformers() && is_scalar($modelData)) {
+        if (!$this->config->getViewTransformers() && !$this->config->getModelTransformers() && !is_bool($modelData) && is_scalar($modelData)) {
```
Comment (HeahDude) [inline src/Symfony/Component/Form/Form.php]: You should replace the scalar check by a string check instead of testing a non-bool.

**Comment by HeahDude** [comment]
Hi @be-bewweb, could you please update to 2.7.10 and confirm your issue before moving on with this PR ?

This bug may have been fixed by #17760.

**Comment by xabbuh** [comment]
ping @be-bewweb :)

**Comment by stof** [comment]
Closing this as the bug is in the code written by the user. Booleans cannot be used as array keys in PHP, as I said above.

**Pull request closed without merging**